### PR TITLE
FIX-#4273: Fix waiting of None-returning methods in benchmark mode

### LINKED ISSUE #4273: BUG: `AttributeError` during using `Series.isin` in benchmark mode
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **Modin version** (`modin.__version__`):  e5e9634357e60925a5a70e56a1d4882d269f533a
- **Python version**: 3.8.11
- **Code we can use to reproduce**:

```python
import numpy as np
import modin.pandas as pd
import modin.config as cfg
cfg.BenchmarkMode.put(True)

nrows = 10

d = ["A", "B", "C", "D"]

s1 = pd.Series([d[np.random.randint(0, 4)] for _ in range(nrows)])
s2 = pd.Series(d)
mask = s2.isin(s1)

print(mask)

```

### Describe the problem

`AttributeError` is appeared during using `Series.isin` in benchmark mode only. 

### Source code / logs

<details><summary>Traceback</summary>

```bash
Traceback (most recent call last):                                                                      File "test.py", line 12, in <module>
    mask = s2.isin(s1)
  File "/localdisk/aprutsko/modin/modin/pandas/base.py", line 1532, in isin                               query_compiler=self._query_compiler.isin(values=values)
  File "/localdisk/aprutsko/modin/modin/core/dataframe/algebra/map.py", line 47, in caller
    query_compiler._modin_frame.map(                                                                    File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 114, in run_f_on_minimally_updated_metadata
    result = f(self, *args, **kwargs)                                                                   File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1514, in map                                                                                                        new_partitions = self._partition_mgr_cls.map_partitions(self._partitions, func)
  File "/localdisk/aprutsko/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py", line 57, in magic
    result_parts = f(*args, **kwargs)
  File "/localdisk/aprutsko/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py", line 280, in map_partitions
    return super(PandasOnRayDataframePartitionManager, cls).map_partitions(
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 58, in wait
    result = func(*args, **kwargs)
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 487, in map_partitions
    preprocessed_map_func = cls.preprocess_func(map_func)
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 114, in preprocess_func
    return cls._partition_class.preprocess_func(map_func)
  File "/localdisk/aprutsko/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py", line 246, in preprocess_func
    return ray.put(func)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/worker.py", line 1795, in put
    object_ref = worker.put_object(
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/worker.py", line 294, in put_object
    serialized_value = self.get_serialization_context().serialize(value)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/serialization.py", line 367, in serialize
    return self._serialize_to_msgpack(value)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/serialization.py", line 347, in _serialize_to_msgpack
    self._serialize_to_pickle5(metadata, python_objects)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/serialization.py", line 307, in _serialize_to_pickle5
    raise e
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/serialization.py", line 303, in _serialize_to_pickle5
    inband = pickle.dumps(
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 620, in dump
    return Pickler.dump(self, obj)
  File "/localdisk/aprutsko/modin/modin/pandas/series.py", line 2550, in __reduce__
    self._query_compiler.finalize()
  File "/localdisk/aprutsko/modin/modin/core/storage_formats/pandas/query_compiler.py", line 254, in finalize
    self._modin_frame.finalize()
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2817, in finalize
    self._partition_mgr_cls.finalize(self._partitions)
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 66, in wait
    all(map(lambda partition: partition.wait() or True, partitions.flatten()))
AttributeError: 'NoneType' object has no attribute 'flatten'
```
</details>

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4273 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release 

The PR modifies `PartitionManager.finalize` method to return finalized partitions as it was agreed in [this thread](https://github.com/modin-project/modin/pull/4404#discussion_r861964252). It also changes the [`@wait_computations_if_benchmark_mode`](https://github.com/modin-project/modin/blob/e3723a63ad06eba96e1ed1e88c1821df5a762cee/modin/core/dataframe/pandas/partitioning/partition_manager.py#L35) decorator to raise a proper error in case of similar error occure again.

### CODE REVIEW

File: docs/release_notes/release_notes-0.15.0.rst [hunk 1]
```diff
@@ -13,6 +13,7 @@ Key Features and Updates
   * FIX-#4385: Get rid of `use-deprecated` option in `pip` (#4386)
   * FIX-#3527: Fix parquet partitioning issue causing negative row length partitions (#4368)
   * FIX-#4330: Override the memory limit to start ray 1.11.0 on Macs (#4335)
+  * FIX-#4273: Fix waiting of None-returning methods in benchmark mode (#4403)
```
Comment (dchigarev) [inline docs/release_notes/release_notes-0.15.0.rst:16]:
```suggestion
  * FIX-#4273: Fix waiting of None-returning methods in benchmark mode (#4404)
```

**Review by anmyachev — APPROVED** [review]
LGTM!

**Comment by anmyachev** [comment]
@dchigarev could you update the PR?

File: modin/test/test_utils.py [hunk 2]
```diff
@@ -251,7 +251,7 @@ def warns_that_defaulting_to_pandas():
         A WarningsChecker checking for a UserWarning saying that Modin is
         defaulting to Pandas.
     """
-    return pytest.warns(UserWarning, match="defaulting to pandas")
+    return pytest.warns(UserWarning, match="[Dd]efaulting to pandas")
```
Comment (dchigarev) [inline modin/test/test_utils.py:254]: some of the warnings (e.g. the ones that raised in `df.__reduce__`) starts with a capital "D", so changed the regexp to match both of the cases

**Comment by dchigarev** [comment]
blocked by #4423 for now

File: modin/pandas/test/internals/test_benchmark_mode.py [hunk 3]
```diff
@@ -15,17 +15,38 @@
...
-    with (
-        warns_that_defaulting_to_pandas()
-        if StorageFormat.get() == "Omnisci"
-        else nullcontext()
-    ):
+    with default_to_pandas_context:
         pd.DataFrame(test_data_values[0]).mean()
+
+
+def test_serialization():
```
Comment (dchigarev) [inline modin/pandas/test/internals/test_benchmark_mode.py:37]: the test doesn't explicitly follow the repro from #4273 but still triggers the same scenario: the original issue occurred due to serialization of Series argument when sending it between workers (so basically doing `series.__reduce__`), which triggered to execute `series.finalize()` that failed in benchmark mode.

the original reproducer couldn't be applied as `Series.isin` itself doesn't work at all with Dask engine (doesn't matter what's the value of `BenchmarkMode` variable is); raised a separate issue for this #4432

File: modin/pandas/test/internals/test_benchmark_mode.py [hunk 4]
```diff
@@ -15,17 +15,38 @@
...
+    assert BenchmarkMode.get()
+
+    sr1 = pd.Series(range(10))
+    # On Omnisci `__finalize__` raises a warning
+    with default_to_pandas_context:
+        constructor, args = sr1.__reduce__()
+    sr2 = constructor(*args)
```
Comment (anmyachev) [inline modin/pandas/test/internals/test_benchmark_mode.py:44]: It's a bit non-obvious, why would need to use these methods?

File: modin/pandas/test/internals/test_benchmark_mode.py [hunk 5]
```diff
@@ -15,17 +15,38 @@
...
-        else nullcontext()
-    ):
+    with default_to_pandas_context:
         pd.DataFrame(test_data_values[0]).mean()
+
+
+def test_serialization():
+    assert BenchmarkMode.get()
+
+    sr1 = pd.Series(range(10))
```
Comment (anmyachev) [inline modin/pandas/test/internals/test_benchmark_mode.py:40]: Maybe parametrize data via pytest?

File: modin/pandas/test/internals/test_benchmark_mode.py (hunk 4 above)
Comment (dchigarev) [reply modin/pandas/test/internals/test_benchmark_mode.py:44]: it's `__reduce__` signature, it returns a tuple of a construction function and an arguments to pass in it in order to deserialize the object
https://github.com/modin-project/modin/blob/8adf28f62c09d49ebb677cf14ad54a3361329b22/modin/pandas/series.py#L2516-L2520

File: modin/pandas/test/internals/test_benchmark_mode.py (hunk 5 above)
Comment (dchigarev) [reply modin/pandas/test/internals/test_benchmark_mode.py:40]: what for? the test doesn't depend on the data provided, it always works the same

File: modin/pandas/test/internals/test_benchmark_mode.py [hunk 6]
```diff
@@ -15,17 +15,38 @@
...
+        constructor, args = sr1.__reduce__()
+    sr2 = constructor(*args)
+    df_equals(sr1, sr2)
+
+    df1 = pd.DataFrame(range(10))
+    # On Omnisci `__finalize__` raises a warning
+    with default_to_pandas_context:
+        constructor, args = df1.__reduce__()
+    df2 = constructor(*args)
+    df_equals(df1, df2)
```
Comment (anmyachev) [inline modin/pandas/test/internals/test_benchmark_mode.py:52]:
```suggestion
@pytest.mark.parametrize("data_type", [pd.Series, pd.DataFrame])
def test_serialization(data_type):
    assert BenchmarkMode.get()

    dataset1 = data_type(range(10))
    # On Omnisci `__finalize__` raises a warning
    with default_to_pandas_context:
        constructor, args = dataset1.__reduce__()
    dataset2 = constructor(*args)
    df_equals(dataset1, dataset2)
```

File: modin/pandas/test/internals/test_benchmark_mode.py (hunk 5 above)
Comment (anmyachev) [reply modin/pandas/test/internals/test_benchmark_mode.py:40]: [Suggestion.](https://github.com/modin-project/modin/pull/4404/files#r861714759)

File: modin/pandas/test/internals/test_benchmark_mode.py (hunk 3 above)
Comment (anmyachev) [reply modin/pandas/test/internals/test_benchmark_mode.py:37]: Why not just explicitly call `finalize` function?

Comment (dchigarev) [reply modin/pandas/test/internals/test_benchmark_mode.py]: added explicit `finalize` test as well

**Review by anmyachev — DISMISSED** [review]
LGTM!

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 7]
```diff
@@ -1273,7 +1275,6 @@ def binary_operation(cls, axis, left, func, right):
         return result if axis else result.T
 
     @classmethod
-    @wait_computations_if_benchmark_mode
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1276]: Why is this removed?

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1276]: Don't the lines added above resolve the issue?

Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1276]: both changes should be made:
1. We shouldn't wrap None-returning methods with this decorator, as they do not return any waitable objects
2. If we did an error and wrapped None-returning method, we still shouldn't fail but just do a return without waiting anything

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1276]: What if we want to measure the following case when BenchmarkMode is enabled.

```python
t0 = time()
pickle.dumps(df) # __reduce__ -> finalize
t1 = time()
print(t1 - t0)
```
Would that give correct timing?

File: docs/release_notes/release_notes-0.15.0.rst [hunk 8]
```diff
@@ -13,6 +13,7 @@ Key Features and Updates
   * FIX-#4385: Get rid of `use-deprecated` option in `pip` (#4386)
   * FIX-#3527: Fix parquet partitioning issue causing negative row length partitions (#4368)
   * FIX-#4330: Override the memory limit to start ray 1.11.0 on Macs (#4335)
+  * FIX-#4273: Fix waiting of None-returning methods in benchmark mode (#4404)
```
Comment (vnlitvinov) [inline docs/release_notes/release_notes-0.15.0.rst:16]: This is inconsistent with PR title, why?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 9]
```diff
@@ -56,6 +56,8 @@ def wait_computations_if_benchmark_mode(func):
         def wait(*args, **kwargs):
             """Wait for computation results."""
             result = func(*args, **kwargs)
+            if result is None:
+                return result
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:60]: I would rather this be an `assert result is not None` or something similar... we shouldn't hide errors

File: modin/pandas/test/internals/test_benchmark_mode.py [hunk 10]
```diff
@@ -13,19 +13,47 @@
...
+from modin.pandas.test.utils import test_data_values, df_equals
 from modin.config import BenchmarkMode, StorageFormat
 from modin.test.test_utils import warns_that_defaulting_to_pandas
 
 
+default_to_pandas_context = (
+    warns_that_defaulting_to_pandas()
+    if StorageFormat.get() == "Omnisci"
+    else nullcontext()
+)
```
Comment (vnlitvinov) [inline modin/pandas/test/internals/test_benchmark_mode.py:29]: This could (potentially) be wrong, as `StorageFormat` could be changed and a test could be called multiple times.
To avoid copy-pasting and retain the re-evaluation of `StorageFormat` you can wrap this in a function.

File: docs/release_notes/release_notes-0.15.0.rst (hunk 8 above)
Comment (RehanSD) [reply docs/release_notes/release_notes-0.15.0.rst:16]: I agree - could we potentially change this PR's title to be more descriptive? Perhaps mention that it's a fix for benchmark mode?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (RehanSD) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1276]: This will no longer be correct, since drain call queue isn't blocking. You'd have to add something that waits on each partition, either by modifying `finalize` to check if benchmark mode is on and doing a wait, or by modifying finalize to return the partitions (perhaps always, perhaps only if benchmark mode is on) and waiting on them.

**Review by RehanSD — CHANGES_REQUESTED** [review]
The fix to the benchmark mode wrapper looks good, but I think a bug may have been introduced to `finalize`.

File: modin/pandas/test/internals/test_benchmark_mode.py (hunk 10 above)
Comment (dchigarev) [reply modin/pandas/test/internals/test_benchmark_mode.py]: Done, wrapped this check into a function and placed it in [`modin/pandas/test/utils.py`](https://github.com/dchigarev/modin/blob/539551e20faf933d382dc11d432ae1638f27273a/modin/pandas/test/utils.py#L1410)

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 9 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: Done, now the decorator is raising an internal error if it failed to extract partitions to wait [[1]](https://github.com/dchigarev/modin/blob/539551e20faf933d382dc11d432ae1638f27273a/modin/core/dataframe/pandas/partitioning/partition_manager.py#L84-L87)

File: docs/release_notes/release_notes-0.15.0.rst (hunk 8 above)
Comment (dchigarev) [reply docs/release_notes/release_notes-0.15.0.rst:16]: renamed the PR to match with the name in release notes

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: Fixed this, now the decorator also seeks for partitions to await inside the function's arguments if the function returns None

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > This will no longer be correct, since drain call queue isn't blocking. You'd have to add something that waits on each partition, either by modifying `finalize` to check if benchmark mode is on and doing a wait, or by modifying finalize to return the partitions (perhaps always, perhaps only if benchmark mode is on) and waiting on them.

It turns out that there is a mismatch with what is written in the docstring of `finalize` method. So we should add a loop where method `wait` will be explicitly called on each partition.

In this case, it will be sufficient to simply remove the decorator (without making changes to it).

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: What do you mean exactly? The docstring says "Perform all deferred calls on partitions.", but not "Perform all deferred calls on partitions and wait until all computations are complete."

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: The current implementation should be understood as: "Trigger all deferred computations on partitions". 

To write "perform" in the docstring, the function should wait for the results. Isn't that what the function is for?

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: Nope, the method is meant to perform all deferred calls, which have been added to the call queues, on partitions in asynchronous way.

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: if so, the simplest way here is return 2d ndarray with partitions

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: Even though this is the simplest way to fix the issue, I think this is the right way and the partition manager should always return updated partitions to construct a new core Modin dataframe to avoid any further side effects.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 11]
```diff
@@ -52,18 +55,40 @@ def wait_computations_if_benchmark_mode(func):
...
+            if hasattr(obj, "wait"):
+                return (obj,)
+            # fastpath for NumPy array of partitions
+            elif (
+                isinstance(obj, np.ndarray)
+                and obj.dtype == np.object
+                and len(obj) > 0
+                and hasattr(obj[np.unravel_index(0, shape=obj.shape)], "wait")
+            ):
+                return obj.flatten()
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:69]: we actually don't care about the type of the returned collection, so just changed docstring to say "Return an iterable..."

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: cc @dchigarev

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 12]
```diff
@@ -52,18 +55,40 @@ def wait_computations_if_benchmark_mode(func):
     """
     if BenchmarkMode.get():
 
+        def extract_waitable(obj):
+            """Return a tuple of waitable objects contained in the `obj`."""
+            if hasattr(obj, "wait"):
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:60]: when the partitions container (`obj`) is not a numpy array the code skips the fast-path and falls down to the inspecting each element of the collection on whether it's waitable or not

for example, this code will be hitting this `hasattr` check for every single element in the list:
```python
extract_waitable([[Waitable(), Waitable()]])
```

added a test case to cover this if-statement as well

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: done, now `PartitionManager.finalize` method returns passed partitions after draining call queues on them

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: What I mean is we should also construct a new core Modin dataframe, as well as update inplace the high-layer DataFrame.

Also, I wonder whether we need the new logic for the decorator as it will hide similar errors to one be resolved in this PR? Because we assume the partition manager should always return updated partitions.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 13]
```diff
@@ -52,18 +55,40 @@ def wait_computations_if_benchmark_mode(func):
...
+
         @wraps(func)
         def wait(*args, **kwargs):
             """Wait for computation results."""
             result = func(*args, **kwargs)
-            if isinstance(result, tuple):
-                partitions = result[0]
+            if result is None:
+                # the operation was applied on partitions inplace so extracting those from arguments
+                partitions = extract_waitable((args, kwargs.values()))
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:81]: Why is this flexibility needed? Can we require a partition manager that any of its functions (where applicable) always return an array of partitions?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > What I mean is we should also construct a new core Modin dataframe, as well as update inplace the high-layer DataFrame.

Not sure about that. What `finalize` is intended to do is to ["Finalize <b><i>CONSTRUCTING</b></i> the dataframe calling all deferred functions which were used to build it"](https://github.com/modin-project/modin/blob/270fd697a82d051f0ab2c54b70b74296699bdcf2/modin/core/storage_formats/base/query_compiler.py#L275). The `finalize` method when called against a finalizing object, have to mutate the actual object, without creating a new one that is considered to be finalized.

> I wonder whether we need the new logic for the decorator as it will hide similar errors to one be resolved in this PR? Because we assume the partition manager should always return updated partitions.

It won't hide similar errors, it will just expand a set of use-cases when the decorator could be applied (now it's possible to use it for inplace methods as well). For the similar-error cases there's now a check, that raises a proper error if so happened. [[1]](https://github.com/dchigarev/modin/blob/2eea67dc78353f5fccfd93d7ba498d83eca9173c/modin/core/dataframe/pandas/partitioning/partition_manager.py#L84-L87)

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 13 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:81]: It's needed just to extend a set of use cases. As we discover errors (inplace method was wrapped by the decorator) only in runtime it's better to be oversecured.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 14]
```diff
@@ -52,18 +55,40 @@ def wait_computations_if_benchmark_mode(func):
...
+        def extract_waitable(obj):
+            """Return a tuple of waitable objects contained in the `obj`."""
+            if hasattr(obj, "wait"):
+                return (obj,)
+            # fastpath for NumPy array of partitions
+            elif (
+                isinstance(obj, np.ndarray)
+                and obj.dtype == np.object
+                and len(obj) > 0
+                and hasattr(obj[np.unravel_index(0, shape=obj.shape)], "wait")
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:67]: Imagine the decorator being applied to the method like this, then we have to distinguish between an actual 2D array with partitions and the metadata one (there is no such cases yet, just being oversafe)
```python
def partition_manager_method():
    return np.array([["some"], ["metadata"]]), np.array([[part1], [part2]])
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 13 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:81]: if now we do not use it, then in this case it is enough to throw an exception. Writing code for the future is not a good idea.

Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: Okay, removed support for inplace functions in the decorator

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: @dchigarev thanks :)

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > BTW, notice that even `PartitionManager.finalize` does not create new finalized partitions, but just returns exactly the same numpy array that was passed as `partitions`.

Please note, the content of returned partitions may become different so I was wondering a case when metadata might be unsynced. However, because we take care enough of syncing metadata at the core Dataframe layer, there shouldn't be such a case and the flow will be correct until partitions (NumPy array) object is immutable. So I think we should probably be explicit and assign new returned partitions to `_partition` field of the core Dataframe. Does this make sense?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 15]
```diff
@@ -19,9 +19,12 @@
 
 from abc import ABC
 from functools import wraps
+from itertools import chain
 import numpy as np
 import pandas
+
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:25]:
```suggestion
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 16]
```diff
@@ -1282,8 +1303,17 @@ def finalize(cls, partitions):
         ----------
         partitions : np.ndarray
             Partitions of Modin Dataframe on which all deferred calls should be performed.
+
+        Returns
+        -------
+        np.ndarray
+            Partitions with the no deferred calls. Note that this is the same partition objects
+            that were passed as `partitions` argument. The value is returned for the sake of
+            interface alignment with other manager's methods and may be safely omitted if performed
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1312]: Is the last sentence still true?

File: modin/pandas/test/utils.py [hunk 17]
```diff
@@ -1403,3 +1405,23 @@ def dict_equals(dict1, dict2):
     for key1, key2 in itertools.zip_longest(sorted(dict1), sorted(dict2)):
         value_equals(key1, key2)
         value_equals(dict1[key1], dict2[key2])
+
+
+def catch_default_to_pandas_warnings_if_omnisci():
+    """
+    Return a StorageFormat-dependent default-to-pandas warnings catching context.
+
+    If StorageFormat is Omnisci, returns a context that asserts for a warning to be
```
Comment (YarShev) [inline modin/pandas/test/utils.py:1414]:
```suggestion
    If StorageFormat is OmniSci, returns a context that asserts for a warning to be
```

File: modin/pandas/test/utils.py [hunk 18]
```diff
@@ -1403,3 +1405,23 @@ def dict_equals(dict1, dict2):
...
+    raised inside it, otherwise returns ``nullcontext``.
+
+    Notes
+    -----
+    As Omnisci execution tends to raise a lot of warnings, it's convenient in execution-common
+    tests to wrap by this context a piece of code that defaults-to-pandas on Omnisci and never
```
Comment (YarShev) [inline modin/pandas/test/utils.py:1420]:
```suggestion
    tests to wrap by this context a piece of code that defaults-to-pandas on OmniSci and never
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 19]
```diff
@@ -52,18 +55,36 @@ def wait_computations_if_benchmark_mode(func):
...
             """Wait for computation results."""
             result = func(*args, **kwargs)
-            if isinstance(result, tuple):
-                partitions = result[0]
-            else:
-                partitions = result
+            partitions = extract_waitable(result)
+            ErrorMessage.catch_bugs_and_request_email(
+                failure_condition=len(partitions) == 0,
+                extra_log=f"No partitions to wait was returned from the wrapped function: {func}",
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:82]:
```suggestion
                extra_log=f"No partition to wait was returned from the wrapped function: {func}",
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:69]: ok

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:67]: This is a bit complex but I think we can leave as is.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:60]: ok

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > So I think we should probably be explicit and assign new returned partitions to _partition field of the core Dataframe. Does this make sense?

Don't know. I'm a bit tensed about this approach as well as about returning "finalized" partitions from `PartitionManager.finalize`. IMHO, assigning new `._partitions` is redundant and would confuse developers as the line literally means `self._partitions = self._partitions` but creates an impression of newly created "finalized" array of partitions. I don't want to be a blocker here as the discussion we had in the PR is already quite big, but I still not really like the proposed approach because of its counterintuitivity.

So would like to hear other opinions from @modin-project/modin-core about this thread and I will just agree with the majority.

The questions are:
1. [`PartitionManager.finalize`](https://github.com/modin-project/modin/blob/270fd697a82d051f0ab2c54b70b74296699bdcf2/modin/core/dataframe/pandas/partitioning/partition_manager.py#L1275-L1286) modifies the passed partition's state by calling the deferred calls on them (no new partitions produced during this), is this okay to return the passed partitions in the result for the sake of interface aligning with other manager's method? My concern is that it creates a false impression of the function's user that the returned array is a new one.
    ```python
    @classmethod
    def finalize(cls, partitions):
        [partition.drain_call_queue() for partition in partitions.flaten()]
        return partitions # return exactly the same object that was passed into the func
    ```
    (we can fix the actual issue (#4273) without that by just inspecting the wrapped method's argument for partitions to wait)
2. If we decide to return partitions from `PartitionManager.finalize`, should we then update `._partitions` with the new ones (actually exactly the same objects) that were returned from `PartitionManager.finalize`? My concern is that it could be really misleading, as the statement would literaly mean `self._partitions = self._partitions`:
    ```python
    def finalize(self):
        self._partitions = self._partition_mgr_class.finalize(self._partitions)
    ```

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > the line literally means `self._partitions = self._partitions`

As I said, this will be the same until `partitions` (NumPy array) object is immutable.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 20]
```diff
@@ -52,18 +54,36 @@ def wait_computations_if_benchmark_mode(func):
...
+            partitions = extract_waitable(result)
+            ErrorMessage.catch_bugs_and_request_email(
+                failure_condition=len(partitions) == 0,
+                extra_log=f"No partition to wait was returned from the wrapped function: {func}",
+            )
             # need to go through all the values of the map iterator
             # since `wait` does not return anything, we need to explicitly add
             # the return `True` value from the lambda
-            all(map(lambda partition: partition.wait() or True, partitions.flatten()))
+            all(map(lambda partition: partition.wait() or True, partitions))
```
Comment (mvashishtha) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:86]: While you're modifying this line, could you please simplify the code to 

```suggestion
            [partition.wait() for partition in partitions]
```

I think the existing line is effectively doing the same thing.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 7 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: @YarShev what if `finalize` function does not return anything, as before? Just add a mode check and call `wait` method on all partitions if necessary. I don't see any problems that in some cases the partition manager does not return anything. 

It seems to me that we can create more situations with possible side effects (for example, sharing an array with several dataframes) if we return an array here.

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > @YarShev what if `finalize` function does not return anything, as before? Just add a mode check and call `wait` method on all partitions if necessary. I don't see any problems that in some cases the partition manager does not return anything.

@anmyachev, imagine a case when the child core Dataframe contains partitions in an immutable object (or we would decide that for the `PandasDataframe`). If we do not return partitions and not assign them to the respective field, the `PandasDataframe` or its child will contain unchanged partitions.

> It seems to me that we can create more situations with possible side effects (for example, sharing an array with several dataframes) if we return an array here.

Not sure I got you. The array will contain updated partitions.

This discussion is really huge. I wish we converged to a conclusion.

cc @modin-project/modin-core

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: I imagined, but in this case we need to change the data structure, as numpy array is mutable. Such a drastic change looks beyond the scope of this pull request.

As long as partitions object is mutable, by returning them we can assign them to another dataframe, which will lead to a situation where changes in one object will affect another.

I wish the same :)

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: > I imagined, but in this case we need to change the data structure, as numpy array is mutable. Such a drastic change looks beyond the scope of this pull request.

Of course, I am saying about the case when the container for partitions is immutable. We do not want to change it now and probably in future for the `PandasDataframe`.

> As long as partitions object is mutable, by returning them we can assign them to another dataframe, which will lead to a situation where changes in one object will affect another.

We would assign partitions to the `self` dataframe, but not to another.

@devin-petersohn, @RehanSD, @vnlitvinov, thoughts?

**Comment by YarShev** [comment]
@anmyachev, should we close this PR in favor of #5006?

**Comment by anmyachev** [comment]
@YarShev yes, good catch!

**Comment by anmyachev** [comment]
@dchigarev feel free to reopen if needed
